### ISSUE TITLE
[lx] LX Needs Incredibly nonstandard and seemingly useless : timerfd_create

### ISSUE THREAD

**Issue by ebuk** [joyent/illumos-joyent#95]
LX Zones attempting to execute a timerfd_create(CLOCK_MONOTONIC, 0)   will receive EPERM.

Linux man: http://man7.org/linux/man-pages/man2/timerfd_create.2.html

Strace output:

````
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
open("/proc/meminfo", O_RDONLY)         = 10
read(10, "MemTotal:  16777216 kB\nMemFree: "..., 8191) = 322
sysinfo({uptime=515187, loads=[31, 112, 342] totalram=4194304, freeram=4173346, sharedram=0, bufferram=0} totalswap=8388608, freeswap=8380392, procs=32}) = 0
close(10)                               = 0
read(3, "V\223\21{@\201\\\272J\327\342c\340Nt[", 16) = 16
gettid()                                = 22033
timerfd_create(CLOCK_MONOTONIC, 0)      = -1 EPERM (Operation not permitted)
futex(0x7ffffc815af0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
recvfrom(14, "\0", 2048, 0, NULL, NULL) = 1
close(10)                               = 0
close(11) 
````

**Comment by rzezeski** [comment]
This issue comes down to the fact that a non-root user doesn't have the `PRIV_PROC_CLOCK_HIGHRES` privilege. 

```
[dbuser@c66lx-prod ~]$ /native/usr/bin/ppriv -D -e ./test-timerfd-create
test-timerfd-cre[13287]: missing privilege "proc_clock_highres" (euid = 500, syscall = 54) needed at secpolicy_clock_highres+0x20
timerfd_create: Operation not permitted

Message from syslogd@c66lx-prod at Jan 22 14:28:33 ...
 kernel:genunix: [ID 702911 kern.notice] test-timerfd-cre[13287]: missing privilege "proc_clock_highres" (euid = 500, syscall = 54) needed at secpolicy_clock_highres+0x20
```

So it would seem like this is an open and shut case. That the administrator should give this priv to the user that needs it. But I think there's a bit more nuance. If the goal is to make lx feel like stock Linux then perhaps there shouldn't be a restriction like this on `timerfd_create(2)`. Adding to the confusion, lx puts no such restriction on `timer_create(2)`, even though these system calls are siblings. The reason for this is that `timer_create(2)` support was added long ago; and the decision back then was to simply lie and map the monotonic clock to the realclock.

```C
/*
 * Since the Illumos CLOCK_HIGHRES clock requires elevated privs, which can
 * lead to a DOS, we use the only other option (CLOCK_REALTIME) when given
 * LX_CLOCK_MONOTONIC.
 */
static int ltos_timer[] = {
        CLOCK_REALTIME,
        CLOCK_REALTIME,
        CLOCK_THREAD_CPUTIME_ID,        /* XXX thread, not process but fails */
        CLOCK_THREAD_CPUTIME_ID,
        CLOCK_REALTIME,
        CLOCK_REALTIME,
        CLOCK_REALTIME
};
...
/*
 * The Illumos timer_create man page says it accepts the following clocks:
 *   CLOCK_REALTIME (3) wall clock
 *   CLOCK_VIRTUAL (1)  user CPU usage clock - No Backend
 *   CLOCK_PROF (2) user and system CPU usage clock - No Backend
 *   CLOCK_HIGHRES (4)  non-adjustable, high-resolution clock
 * However, in reality the Illumos timer_create only accepts CLOCK_REALTIME
 * and CLOCK_HIGHRES, and since we can't use CLOCK_HIGHRES in a zone, we're
 * down to one clock.
 */
long
lx_timer_create(int clock, struct sigevent *lx_sevp, timer_t *tid)
```

(The comment about using `CLOCK_HIGHRES` in a zone is no longer true.)

My naive guess is that the priv is here to protect the cyclic subsystem. E.g., if someone registers a bunch of really tight intervals then it would eat all the CPUs time while they service interrupts (note, I'm making a guess here, my point is that there was _some_ reason to initially add this priv).

I wonder if we could add some logic to the lx `timerfd_create` that allows unrestricted use of the monotonic clock but also prevents the user from being malicious. E.g., perhaps the size of the interval could have a lower bound?

**Comment by rzezeski** [comment]
One thing I forgot to mention. If the highres clock timer is a potential DOS vector, then should the zone have access to it at all? Should the root user of a zone have power to DOS the hypervisor?

**Comment by jjelinek** [comment]
Filed OS-5263 for this.

**Comment by jjelinek** [comment]
Fix in 873e369
